### PR TITLE
[mono][swift-interop] Support for Swift struct lowering in direct P/Invoke returns

### PR DESCRIPTION
**Pull request by matouskozak**
## Description
This PR adds support in Mono runtime for Swift `@frozen` struct lowering in returns. The scope is limited to direct P/Invoke scenario with reverse P/Invoke following up in a separate PR. The changes cover support over all Mono supported backend engines: miniJIT, interpreter and AOT.

In Swift, structures can be lowered into <= 4 primitives that are subsequently returned via registers. This is a simplified scenario in comparison to lowering of arguments where lowered elements can also be passed on stack when maximum number of registers used for arguments is exceeded.

## Implementation
The lowering support for returns is handled at the codegen level (`mini-<arch>.c`) by introducing new storage type (`ArgSwiftVtypeLoweredRet`/`ArgSwiftValuetypeLoweredRet`) that represents a structure that is returned via combination of integer and float registers. 

The first step of struct lowering is handled in `get_call_info` -> `add_return_valuetype_swiftcall` where struct lowering is computed. In scenarios where struct can be lowered (fits in <= 4 registers) the newly introduced storage type is used, otherwise the struct is returned via reference.

1. Returned by lowered sequence:
    - The re-composition of struct is handled in `emit_move_return_value` where the values stored in return registers are loaded into prepared memory at specific offsets.

2. Returned by reference:
    - On arm64, handled as regular `ArgVtypeByRef` with no extra changes.
    - On x64, handled as `ArgValuetypeAddrInIReg` with specific handling where the reference is temporarily hold in `R10` reg. and moved to `RAX` reg. (`amd64_handle_swift_return_buffer_reg`) before emitting the call to native code. That's because Swift expects that the return buffer reference is passed in `RAX` reg. but in Mono runtime, this reg. serves multiple purposes and we cannot easily prevent `RAX` clobbering by other parts of code.

## Verification
The changes are verified using the `Interop/Swift/SwiftRetAbiStress` runtime tests. The extended set of 5000 test cases was used locally to further verify the lowering support. The fullAOT functionality was only verified in local settings due to missing coverage on CI.

## Other changes
- refactoring of argument Swift `@frozen` struct lowering in `method-to-ir.c`
- refactoring of Swift return buffer handling (`amd64_handle_swift_return_buffer_reg`)
- adding more `#ifdef MONO_ARCH_HAVE_SWIFTCALL` to better separate swiftcall support
- add missing `ArgSwiftError` entry in `mono_arch_get_llvm_call_info` on arm64. This is just a placeholder because we don't compile P/Invoke wrappers by LLVM. However, on arm64, the missing entry was causing failures in AOT-llvm compilation. 

---

Contributes to https://github.com/dotnet/runtime/issues/102077

### CODE REVIEW

File: src/mono/mono/mini/tramp-amd64.c [hunk 1]
```diff
@@ -1179,7 +1179,7 @@ mono_arch_get_interp_to_native_trampoline (MonoTrampInfo **info)
 
 	/* save all return floating registers in the CallContext */
 	for (i = 0; i < FLOAT_RETURN_REGS; i++)
-		amd64_sse_movsd_membase_reg (code, AMD64_R11, MONO_STRUCT_OFFSET (CallContext, fregs) + i * sizeof (double), i);
+		amd64_sse_movsd_membase_reg (code, AMD64_R11, MONO_STRUCT_OFFSET (CallContext, fregs) + float_return_regs [i] * sizeof (double), float_return_regs [i]);
```
Comment (jkurdek) [inline src/mono/mono/mini/tramp-amd64.c:1182]: Why is this change necessary only for amd64?

File: src/mono/mono/mini/mini-arm64.c [hunk 2]
```diff
@@ -1960,10 +2013,40 @@ arg_get_val (CallContext *ccontext, ArgInfo *ainfo, gpointer dest)
...
+			float *dest_float = (float*)dest;
+			for (int k = 0; k < ainfo->nregs; k++) {
+				*dest_float = *(float*)&ccontext->fregs [ainfo->reg + k];
+				dest_float++;
+			}
+			break;
+		}
+#ifdef MONO_ARCH_HAVE_SWIFTCALL
+		case ArgSwiftVtypeLoweredRet: {
+			int i;
```
Comment (jkurdek) [inline src/mono/mono/mini/mini-arm64.c:2027]: NIT: There seems to be some formatting issues here. Tabs vs space I believe.

**Review by jkurdek — APPROVED** [review]
Good job! 

You mentioned you tested these changes on extended set of 5000 tests in `Interop/Swift/SwiftRetAbiStress`. Is there any description somewhere how to get / generate this set?  Could be useful for testing other changes.

File: src/mono/mono/mini/tramp-amd64.c (hunk 1 above)
Comment (matouskozak) [reply src/mono/mono/mini/tramp-amd64.c:1182]: On arm64, the argument and return registers are the same (x0-x7). Whereas on amd64, different registers are used for arguments and returns.

For arm64, this gets handled at https://github.com/dotnet/runtime/blob/4185f94ec85e619b7dfa82357ac853846dfda3af/src/mono/mono/mini/tramp-arm64.c#L773-L779

File: src/mono/mono/mini/mini-amd64.c [hunk 3]
```diff
@@ -623,6 +669,12 @@ add_valuetype (MonoMethodSignature *sig, ArgInfo *ainfo, MonoType *type,
 #ifdef TARGET_WIN32
 	add_valuetype_win64 (sig, ainfo, type, is_return, gr, fr, stack_size);
 #else
+#ifdef MONO_ARCH_HAVE_SWIFTCALL
+	if (is_return && sig->pinvoke && mono_method_signature_has_ext_callconv (sig, MONO_EXT_CALLCONV_SWIFTCALL)) {
+		add_return_valuetype_swiftcall (sig, ainfo, type, gr, fr, stack_size);
```
Comment (kotlarmilos) [inline src/mono/mono/mini/mini-amd64.c:674]: Is it possible to move this to `get_call_info`? This should be invoked only for wrappers and could be simplified without changes in `add_valuetype`.

File: src/mono/mono/mini/method-to-ir.c [hunk 4]
```diff
@@ -7547,34 +7547,27 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
...
 						} else {
 							// For structs that cannot be lowered, we change the argument to byref type
-							ptype = mono_class_get_byref_type (mono_defaults.typed_reference_class);
-							// Load the address of the struct
-							MonoInst *struct_base_address =  mono_compile_create_var (cfg, mono_get_int_type (), OP_LOCAL);
-							CHECK_ARG (idx_param);
-							NEW_ARGLOADA (cfg, struct_base_address, idx_param);
-							MONO_ADD_INS (cfg->cbb, struct_base_address);
 							*sp++ = struct_base_address;
+							ptype = mono_class_get_byref_type (mono_defaults.typed_reference_class);
```
[the remark this answers is not in the document]
Comment (matouskozak) [reply src/mono/mono/mini/method-to-ir.c:7567]: > It is not clear to me why we can use `mono_class_get_byref_type (mono_defaults.typed_reference_class)` instead of `mono_class_get_byref_type (klass);`

I'm not sure either. I did some investigating and it seems that the all byref types get changed to generic `MONO_TYPE_I` which is then handled by codegen as a pointer-sized integer. I will change it to `mono_class_get_byref_type (klass)` as it makes more sense.

File: src/mono/mono/mini/mini-amd64.c (hunk 3 above)
Comment (matouskozak) [reply src/mono/mono/mini/mini-amd64.c]: It's possible, changed!

**Comment by fanyang-mono** [comment]
Is there any follow-up work we need to do for LLVM?

**Comment by matouskozak** [comment]
> Is there any follow-up work we need to do for LLVM?

At the moment no, Mono doesn't use LLVM to compile P/Invoke wrappers so it fallbacks to Mini. I just added the entries inside `mono_arch_get_llvm_call_info` for arm64 to avoid asserting there (not needed on amd64)

File: src/mono/mono/mini/mini-amd64.h [hunk 5]
```diff
@@ -176,6 +176,10 @@ struct sigcontext {
  * reproduceable results for benchmarks */
 #define MONO_ARCH_CODE_ALIGNMENT 32
 
+#if defined(TARGET_OSX) || defined(TARGET_APPLE_MOBILE)
+#define MONO_ARCH_HAVE_SWIFTCALL 1
```
[the remark this answers is not in the document]
Comment (matouskozak) [reply src/mono/mono/mini/mini-amd64.h:180]: I think that it could be set in `src/mono/CMakeLists.txt` by something like:
```
if ((TARGET_AMD64 OR TARGET_ARM64) AND (TARGET_OSX or TARGET_APPLE_MOBILE))
    set(MONO_ARCH_HAVE_SWIFTCALL 1)
endif()
```
However, CoreCLR also does set it in the respective arch files e.g. https://github.com/dotnet/runtime/blob/a5c1c9f364817a1fe2aa306ed33963cb41ab63fc/src/coreclr/jit/targetamd64.h#L563

What do you think is better? (cc: @kotlarmilos, @jkurdek)

[the remark this answers is not in the document]
Comment (kotlarmilos) [reply src/mono/mono/mini/mini-amd64.h:180]: It might be cleaner to move it to `src/mono/CMakeLists.txt`, but don't have a strong opinion about it. Since this is not a new change, I suggest doing it in a follow-up PR.

**Review by kotlarmilos — APPROVED** [review]

**Review by fanyang-mono — APPROVED** [review]

**Review by BrzVlad — APPROVED** [review]
